pkg/alertmanager/state_replication.go

```diff
@@ -134,6 +192,26 @@ func (s *state) Ready() bool {
 	return s.Service.State() == services.Running
 }
 
+// mergeFullStates attempts to merge all full states received from peers during settling.
+func (s *state) mergeFullStates(fs []*clusterpb.FullState) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	for _, f := range fs {
+		for _, p := range f.Parts {
+			level.Debug(s.logger).Log("msg", "merging full state", "user", s.userID, "key", p.Key, "bytes", len(p.Data))
+
+			if st, ok := s.states[p.Key]; !ok {
+				level.Error(s.logger).Log("msg", "key not found while merging full state", "user", s.userID, "key", p.Key)
+			} else {
+				if err := st.Merge(p.Data); err != nil {
```

Comment:
Isn't this a breaking error? Shouldn't we consider the settlement failed in case we can't merge the data (for any reason)?

I would return error from this function in the case, and consider the settlement failed in the caller. This will be important once we'll add the feature to load the state from storage (this is one of the cases where we should fallback to storage, I think).

Reply:
Agreed. Looking at the code for [silences](https://github.com/prometheus/alertmanager/blob/master/silence/silence.go#L821) and [notifications](https://github.com/prometheus/alertmanager/blob/master/nflog/nflog.go#L526), it seems the only cause of this error (currently) is failure to decode the message from the wire. Makes sense to fall back to the persisted state, if available.

I'm actually think if we should do the same for `key not found`?